src/main/java/engine/entities/Board.java

```diff
@@ -71,16 +100,39 @@ public Move getPreviousMove() {
         return null;
     }
 
+    /**
+     * Sets current player to the given player.
+     *
+     * @param player the player whose turn it is to make a move
+     */
+    public void setCurrentPlayer(Player player) {
+        this.currentPlayer = player;
+    }
+
     /**
      * Return the Player object corresponding to
      * the player whose turn it is to make a move.
      *
      * @return the Player object
      */
     public Player getCurrentPlayer() {
-        // TODO
-        // the Board probably needs to know about this too
-        return null;
+        return this.currentPlayer;
+    }
+
+    /**
+     * Sets the current player to the next one in
+     * the players array. If this is the "last" player,
+     * wrap back around to the first one.
+     * If the current player has yet to be set,
+     * set it to the first player in the array.
+     */
+    public void advanceCurrentPlayer() {
```

Comment:
We will consider moving this method into the Game class as Board seems not to need the information of which player's turn it is. The process by which turns are executed is outlined here: https://docs.google.com/document/d/1oo0thyl2tfIQTjTtnxEBpouTDv4sWtX3vEecNI4_wAU/edit

Reply:
I remembered why Board is given the player's turn; this is what ensured that it is possible to advance the player's turn from within a `Move` object's `execute()` method. Indeed, it is better to refactor so that most of the "universal" actions associated with executing a move are handled elsewhere. This is the role filled by `MoveTryer` (see CRC cards). Indeed, this can and should be moved out.

I don't quite see the issue with undoing moves. An undo simply goes back to the previous board. Yes, it reverts to the previous player's turn—it is *the previous player's* move which is being undone. This behaviour is consistent with most two-player chess apps.

The only time one would have to undo multiple moves is if they were playing against a bot (in which case there is no phase in which one awaits the bot's move and has an opportunity to undo). Then it makes sense for the undo to revert two boards back to return to the last turn on which it was the non-bot player's turn.

Reply:
`PostMoveValidator` will need to be based on a Game rather than a Board if we choose this paradigm. Overall, it becomes difficult to work with just Board objects. Much of the program would need to be restructured.

Reply:
Actually, this is false. It is still possible to verify whether a move is valid but it must assume that the piece at a move's origin has the colour of the moving player.